aeron-archive/src/main/java/io/aeron/archive/ArchiveConductor.java

```diff
@@ -1328,54 +1329,86 @@ void migrateSegments(
-                    break;
-                }
+                return;
+            }
 
-                final String dstFile = segmentFileName(dstRecordingId, position);
-                if (!srcFile.renameTo(new File(archiveDir, dstFile)))
-                {
-                    final String msg = "failed to rename " + srcFile + " to " + dstFile;
-                    controlSession.sendErrorResponse(correlationId, msg, controlResponseProxy);
-                    return;
-                }
+            final boolean canPrepend = srcRecordingSummary.stopPosition == dstRecordingSummary.startPosition;
+            final boolean canAppend = srcRecordingSummary.startPosition == dstRecordingSummary.stopPosition;
 
-                attachedSegmentCount++;
-                position -= segmentLength;
+            final long seamPosition;
+
+            if (canPrepend)
+            {
+                seamPosition = srcRecordingSummary.stopPosition;
+            }
+            else if (canAppend)
+            {
+                seamPosition = srcRecordingSummary.startPosition;
+            }
+            else
+            {
+                final String msg = "invalid migrate: src and dst are not contiguous" +
+                    " srcStartPosition=" + srcRecordingSummary.startPosition +
+                    " srcStopPosition=" + srcRecordingSummary.stopPosition +
+                    " dstStartPosition=" + dstRecordingSummary.startPosition +
+                    " dstStopPosition=" + dstRecordingSummary.stopPosition;
+                controlSession.sendErrorResponse(correlationId, msg, controlResponseProxy);
+                return;
+            }
+
+            if (isSeamPosSegmentUnaligned(controlSession, correlationId, "src", srcRecordingSummary, seamPosition) ||
+                isSeamPosSegmentUnaligned(controlSession, correlationId, "dst", dstRecordingSummary, seamPosition))
+            {
+                return;
             }
 
-            if (addDeleteSegmentsSession(correlationId, srcRecordingId, controlSession, files) >= 0)
+            final ArrayDeque<String> emptyFollowingSrcSegment = new ArrayDeque<>();
+
+            final long movedSegmentCount = moveAllSegments(
+                controlSession,
+                correlationId,
+                srcRecordingId,
+                dstRecordingId,
+                srcRecordingSummary,
+                emptyFollowingSrcSegment);
+
+            if (movedSegmentCount >= 0)
             {
-                catalog.startPosition(dstRecordingId, startPosition);
-                catalog.stopPosition(srcRecordingId, startPosition);
-                controlSession.sendOkResponse(correlationId, attachedSegmentCount, controlResponseProxy);
+                final int segmentToBeDeletedCount =
+                    addDeleteSegmentsSession(correlationId, srcRecordingId, controlSession, emptyFollowingSrcSegment);
 
-                if (files.isEmpty())
+                if (segmentToBeDeletedCount >= 0)
```

Comment:
Minor but `toBeDeletedSegmentCount` would be more consistent.

Reply:
:+1: Fixed in latest commits.